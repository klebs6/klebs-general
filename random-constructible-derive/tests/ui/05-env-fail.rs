extern crate random_constructible;
extern crate random_constructible_derive;

use random_constructible::*;
use random_constructible_derive::*;

#[derive(Debug,Default,RandConstruct, Copy, Clone, PartialEq, Eq, Hash)]
enum ProbabilityEnum {
    #[default]
    #[default_unnormalized_construction_probability = 2.0]
    VariantX,
    #[default_unnormalized_construction_probability = 3.0]
    VariantY,
    VariantZ, // Default probability should be 1.0
}

#[derive(Debug,Default,RandConstruct, Copy, Clone, PartialEq, Eq, Hash)]
enum MagicItem {
    #[default]
    #[default_unnormalized_construction_probability = 2.0]
    Cake,
    #[default_unnormalized_construction_probability = 3.0]
    Banana,
    Watermelon, // Default probability should be 1.0
}

#[derive(RandConstructEnvironment)]
pub struct Env;

rand_construct_env!{Env => ProbabilityEnum {
    VariantX       => 1.0,
    VariantY       => 2.0,
}}

fn main() {
    let map = ProbabilityEnum::create_default_probability_map();
    assert_eq!(map.get(&ProbabilityEnum::VariantX), Some(&2.0));
    assert_eq!(map.get(&ProbabilityEnum::VariantY), Some(&3.0));
    assert_eq!(map.get(&ProbabilityEnum::VariantZ), Some(&1.0));

    for _ in 0..3 {
        let x = Env::create_random::<ProbabilityEnum>();
        let y = Env::create_random_uniform::<ProbabilityEnum>();
        assert!(x != ProbabilityEnum::VariantZ);
        assert!(y != ProbabilityEnum::VariantZ);
    }

    // this should not compile because we have not configured Env as a MagicItem provider
    for _ in 0..3 {
        let x = Env::create_random::<MagicItem>();
        let y = Env::create_random_uniform::<MagicItem>();
        assert!(x == MagicItem::Watermelon);
        assert!(y == MagicItem::Watermelon);
    }
}
