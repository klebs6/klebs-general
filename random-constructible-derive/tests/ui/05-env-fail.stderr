error[E0277]: the trait bound `Env: RandConstructProbabilityMapProvider<MagicItem>` is not satisfied
  --> tests/ui/05-env-fail.rs:50:17
   |
50 |         let x = Env::create_random::<MagicItem>();
   |                 ^^^ the trait `RandConstructProbabilityMapProvider<MagicItem>` is not implemented for `Env`
   |
   = help: the trait `RandConstructProbabilityMapProvider<ProbabilityEnum>` is implemented for `Env`
   = help: for that trait implementation, expected `ProbabilityEnum`, found `MagicItem`
note: required by a bound in `create_random`
  --> $CARGO/random-constructible-0.4.0/src/lib.rs
   |
   |     fn create_random<R>() -> R
   |        ------------- required by a bound in this associated function
...
   |         Self: RandConstructProbabilityMapProvider<R> + Sized,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RandConstructEnvironment::create_random`

error[E0277]: the trait bound `Env: RandConstructProbabilityMapProvider<MagicItem>` is not satisfied
  --> tests/ui/05-env-fail.rs:51:17
   |
51 |         let y = Env::create_random_uniform::<MagicItem>();
   |                 ^^^ the trait `RandConstructProbabilityMapProvider<MagicItem>` is not implemented for `Env`
   |
   = help: the trait `RandConstructProbabilityMapProvider<ProbabilityEnum>` is implemented for `Env`
   = help: for that trait implementation, expected `ProbabilityEnum`, found `MagicItem`
note: required by a bound in `create_random_uniform`
  --> $CARGO/random-constructible-0.4.0/src/lib.rs
   |
   |     fn create_random_uniform<R>() -> R
   |        --------------------- required by a bound in this associated function
...
   |         Self: RandConstructProbabilityMapProvider<R> + Sized,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RandConstructEnvironment::create_random_uniform`
